etna/whatson/models.py

```diff
@@ -0,0 +1,621 @@
+                FieldPanel("contact_info"),
+            ],
+            heading=_("booking information"),
+        ),
+    ]
+
+    def clean(self):
+        """
+        Check that the venue address and video conference information are
+        provided for the correct venue type.
+        """
+
+        if self.venue_type:
+            if self.venue_type == VenueType.HYBRID and (
+                not self.venue_address or not self.video_conference_info
+            ):
+                raise ValidationError(
+                    {
+                        "venue_address": _(
+                            "The venue address is required for hybrid events."
+                        ),
+                        "venue_space_name": _(
+                            "The venue space name is required for hybrid events."
+                        ),
+                        "video_conference_info": _(
+                            "The video conference information is required for hybrid events."
+                        ),
+                    }
+                )
+            elif self.venue_type == VenueType.IN_PERSON and not self.venue_address:
+                raise ValidationError(
+                    {
+                        "venue_address": _(
+                            "The venue address is required for in person events."
+                        ),
+                        "venue_space_name": _(
+                            "The venue space name is required for hybrid events."
+                        ),
+                    }
+                )
+            elif self.venue_type == VenueType.ONLINE and not self.video_conference_info:
+                raise ValidationError(
+                    {
+                        "video_conference_info": _(
+                            "The video conference information is required for online events."
+                        ),
+                    }
+                )
+        return super().clean()
+
+    def serializable_data(self):
+        # Keep aggregated field values out of revision content
+
+        data = super().serializable_data()
+
+        for field_name in ("start_date", "end_date"):
+            data.pop(field_name, None)
+
+        return data
+
+    def with_content_json(self, content):
+        """
+        Overrides Page.with_content_json() to ensure page's `start_date` and `end_date`
+        value is always preserved between revisions.
+        """
+        obj = super().with_content_json(content)
+        obj.start_date = self.start_date
+        obj.end_date = self.end_date
+        return obj
+
+    def save(self, *args, **kwargs):
+        """
+        Set the event start date to the earliest session start date.
+        Set the event end date to the latest session end date.
+        """
+        sessions = [session for session in self.sessions.all()]
+
+        self.start_date = sessions[0].start
+        self.end_date = sessions[-1].end
```

Comment:
There's actually nothing in place to enforce that the last item is the one with the latest 'end' value, so we might want to be a bit more cautious here. Maybe something like this would be better in this scenario:

```
min_start = None
max_end = None
for session in self.sessions.all():
    if min_start is None or session.start < min_start:
        min_start = session.start
    if max_end is None or session.end > max_start:
        max_end = session.max_end

self.start_date = min_start
self.end_date = max_end
```